index.bs

```diff
@@ -4648,71 +4656,154 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
+
+            1. [=Set device pixel ratio override=] with |navigable| and |viewport overrides map|[|user context|]["<code>devicePixelRatio</code>""].
+
 <div algorithm="remote end steps for browsingContext.setViewport">
 
 The [=remote end steps=] with |command parameters| are:
 
-1. Let |navigable id| be the value of the <code>context</code> field of |command
-   parameters|.
+1. If |command parameters| [=map/contains=] "<code>userContexts</code>"
+   and |command parameters| [=map/contains=] "<code>context</code>",
+   return [=error=] with [=error code=] [=invalid argument=].
 
-1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
-   |navigable id|.
+1. Let |navigables| be a [=/set=].
 
-1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
-   [=error code=] [=invalid argument=].
+1. If the <code>context</code> field of |command parameters| is present:
 
-1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
-   parameters with the given |command parameters| for any reason, return
-   [=error=] with [=error code=] [=unsupported operation=].
+   1. Let |navigable id| be the value of the <code>context</code> field of |command
+      parameters|.
 
-1. If |command parameters| [=map/contains=] the <code>viewport</code> field:
+   1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
+      |navigable id|.
 
-   1. Let |viewport| be the |command parameters|["<code>viewport</code>"].
+   1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
+      [=error code=] [=invalid argument=].
 
-   1. If |viewport| is not null, set the width of |navigable|'s [=layout
-      viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
-      set the height of the |navigable|'s [=layout viewport=] to be the
-      <code>height</code> of |viewport| in CSS pixels.
+   1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
+      parameters with the given |command parameters| for any reason, return
+      [=error=] with [=error code=] [=unsupported operation=].
 
-   1. Otherwise, set the |navigable|'s [=layout viewport=] to the
-      implementation-defined default.
+   1. [=set/Append=] |navigable| to |navigables|.
 
-1. Run the [[cssom-view-1#resizing-viewports]] steps.
+1. Otherwise, if the <code>userContexts</code> field of |command parameters| is present:
 
-1. If |command parameters| [=map/contains=] the <code>devicePixelRatio</code> field:
+   1. For each |user context id| of |command parameters|["<code>userContexts</code>"]:
 
-   1. Let |device pixel ratio| be the |command
-      parameters|["<code>devicePixelRatio</code>"].
+      1. Set |user context| to [=get user context=] with |user context id|.
 
-   1. For the |navigable| and all [=descendant navigables=]:
+      1. If |user context| is null, return [=error=] with [=error code=] [=no such user context=].
```

Comment:
This failure inside the loop means that we'd partially apply the command to any user contexts before the one that doesn't exist; we should probably fail if any don't exist.